installer/PowerToysSetupCustomActions/CustomAction.cpp

```diff
@@ -571,6 +563,66 @@ UINT __stdcall TelemetryLogRepairFailCA(MSIHANDLE hInstall)
     return WcaFinalize(er);
 }
 
+std::optional<std::wstring> getMsiPackageInstalledPath(const wchar_t* product_upgrade_code, const wchar_t* file_component)
+{
+    constexpr size_t guid_length = 39;
+    wchar_t product_ID[guid_length];
+    if (const bool found = ERROR_SUCCESS == MsiEnumRelatedProductsW(product_upgrade_code, 0, 0, product_ID); !found)
+    {
+        return std::nullopt;
+    }
+
+    if (const bool installed = INSTALLSTATE_DEFAULT == MsiQueryProductStateW(product_ID); !installed)
+    {
+        return std::nullopt;
+    }
+
+    DWORD buf_size = MAX_PATH;
+    wchar_t buf[MAX_PATH];
+    if (ERROR_SUCCESS == MsiGetProductInfoW(product_ID, INSTALLPROPERTY_INSTALLLOCATION, buf, &buf_size) && buf_size)
+    {
+        return buf;
+    }
+
+    DWORD package_path_size = 0;
+
+    if (ERROR_SUCCESS != MsiGetProductInfoW(product_ID, INSTALLPROPERTY_LOCALPACKAGE, nullptr, &package_path_size))
+    {
+        return std::nullopt;
+    }
+    std::wstring package_path(++package_path_size, L'\0');
+
+    if (ERROR_SUCCESS != MsiGetProductInfoW(product_ID, INSTALLPROPERTY_LOCALPACKAGE, package_path.data(), &package_path_size))
+    {
+        return std::nullopt;
+    }
+    package_path.resize(size(package_path) - 1); // trim additional \0 which we got from MsiGetProductInfoW
+
+    wchar_t path[256];
+    DWORD path_size = 256;
+    MsiGetComponentPathW(product_ID, file_component, path, &path_size);
+    if (!path_size)
+    {
+        return std::nullopt;
+    }
+    PathCchRemoveFileSpec(path, path_size);
+    return path;
+}
+
+UINT __stdcall DetectPrevInstallPathCA(MSIHANDLE hInstall)
+{
+    HRESULT hr = S_OK;
+    UINT er = ERROR_SUCCESS;
+    hr = WcaInitialize(hInstall, "DetectPrevInstallPathCA");
+
+    if (auto install_path = getMsiPackageInstalledPath(POWERTOYS_UPGRADE_CODE, POWERTOYS_EXE_COMPONENT))
+    {
+        MsiSetPropertyW(hInstall, L"INSTALLFOLDER", install_path->data());
+    }
+    er = SUCCEEDED(hr) ? ERROR_SUCCESS : ERROR_INSTALL_FAILURE;
+    return WcaFinalize(er);
+}
```

Comment:
I would like to see a nice global try and catch in both of the functions above, we have already too many issues with the MSI, having a custom action that could potentially fail is something we need to avoid at all costs.
This is regardless of the code in the function, even if it was a simple one liner!

Reply:
none of the API methods called in those functions throw, except for the allocation failure due. if we'd have a try/catch for them, what do we want to return? ERROR_SUCCESS anyway?

Reply:
Yes

Reply:
I've wrapped the only place of `getMsiPackageInstalledPath` invocation in try/catch block. Now
`    er = SUCCEEDED(hr) ? ERROR_SUCCESS : ERROR_INSTALL_FAILURE;` only depends on the `WcaInitialize`